stlab/concurrency/actor.hpp

```diff
@@ -0,0 +1,179 @@
+
+// lambda pack captures are a C++20 feature, but both clang and msvc support it, so we should be
+// okay to use it even under C++17.
+#if __clang__
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wc++20-extensions"
+#endif // __clang__
+
+/**************************************************************************************************/
+
+namespace stlab {
+inline namespace v1 {
+
+/**************************************************************************************************/
+
+namespace detail {
+
+/**************************************************************************************************/
+
+struct temp_thread_name {
+    explicit temp_thread_name(const char* name) { stlab::set_current_thread_name(name); }
+    ~temp_thread_name() { stlab::set_current_thread_name(""); }
+};
+
+/**************************************************************************************************/
+
+template <class T>
+struct value_instance {
+    std::optional<T> _x;
+};
+
+template <>
+struct value_instance<void> {};
+
+/**************************************************************************************************/
+
+template <class T>
+struct actor_instance : public std::enable_shared_from_this<actor_instance<T>> {
+    template <class Executor>
+    explicit actor_instance(Executor&& e, std::string&& name) :
+        _q(std::forward<Executor>(e)), _name(std::move(name)) {}
+
+    template <class... Args>
+    void initialize(Args&&... args) {
+        // We want to construct the object instance in the executor where
+        // it will be running. We cannot initialize in the constructor because
+        // `shared_from_this` will throw `bad_weak_ptr`.
+        _q(
+            [_this = this->shared_from_this()](auto&&... args) mutable {
+                temp_thread_name ttn(_this->_name.c_str());
+                _this->_instance._x = T(std::forward<Args>(args)...);
+            },
+            std::forward<Args>(args)...)
+            .detach();
+    }
+
+    auto set_name(std::string&& name) { _name = std::move(name); }
+
+    template <typename F, typename... Args>
+    auto send(F&& f, Args&&... args) {
+        auto task = [_f = std::forward<F>(f),
+                     _this = this->shared_from_this()](auto&&... args) mutable {
+            temp_thread_name ttn(_this->_name.c_str());
+            if constexpr (std::is_same_v<T, void>) {
+                return std::move(_f)(std::forward<Args>(args)...);
+            } else {
+                return std::move(_f)(*(_this->_instance._x), std::forward<Args>(args)...);
+            }
+        };
+        return _q(std::move(task), std::forward<Args>(args)...);
+    }
+
+    template <typename R, typename F, typename... Args>
+    auto then(stlab::future<R>&& future, F&& f, Args&&... args) {
+        auto task = [_f = std::forward<F>(f), ... _args = std::forward<Args>(args),
+                     _this = this->shared_from_this()](R&& x) mutable {
```

Comment:
x as value and not as r value ref

Reply:
I have removed this routine in a subsequent update to the class.